projects/batfish/src/test/java/org/batfish/job/BatfishJobExecutorTest.java

```diff
@@ -0,0 +1,214 @@
+package org.batfish.job;
+
+import static org.hamcrest.Matchers.lessThanOrEqualTo;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertThat;
+
+import com.google.common.collect.Lists;
+import com.google.common.collect.Sets;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import org.batfish.common.BatfishException;
+import org.batfish.common.BatfishLogger;
+import org.batfish.common.BatfishLogger.BatfishLoggerHistory;
+import org.batfish.config.Settings;
+import org.batfish.datamodel.answers.AnswerElement;
+import org.batfish.job.BatfishJobExecutor.HandleProcessingErrorException;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+
+/** Tests for {@link BatfishJobExecutor}. */
+public class BatfishJobExecutorTest {
+
+  private static final long TEST_JOB_DELAY = 1000L;
+  private static final String TEST_EXECUTOR_DESC = "Test Job Executor";
+
+  @Rule public ExpectedException _thrown = ExpectedException.none();
+  BatfishLogger _logger;
+
+  @Before
+  public void setup() {
+    _logger = new BatfishLogger(BatfishLogger.LEVELSTR_INFO, false);
+  }
+
+  @Test
+  public void testExecuteJobsResults() {
+    Settings settings = new Settings();
+    settings.setHaltOnParseError(false);
+    settings.setSequential(false);
+
+    List<BfTestJob> jobs = new ArrayList<>();
+    jobs.add(new BfTestJob(settings, "result1"));
+    jobs.add(new BfTestJob(settings, "result2"));
+
+    Set<String> output = new HashSet<>();
+    BfTestAnswerElement ae = new BfTestAnswerElement();
+    BatfishJobExecutor<BfTestJob, BfTestAnswerElement, BfTestResult, Set<String>> executor =
+        new BatfishJobExecutor<>(
+            settings, _logger, settings.getHaltOnParseError(), TEST_EXECUTOR_DESC);
+    executor.executeJobs(jobs, output, ae);
+
+    // checking the outputs produced by the tasks
+    assertEquals(output, Sets.newHashSet("result1", "result2"));
+  }
+
+  @Test
+  public void testExecuteJobsTime() {
+    Settings settings = new Settings();
+    settings.setHaltOnParseError(false);
+
+    List<BfTestJob> jobs = new ArrayList<>();
+    jobs.add(new BfTestJob(settings, "result1"));
+    jobs.add(new BfTestJob(settings, "result2"));
+
+    Set<String> output = new HashSet<>();
+    BfTestAnswerElement ae = new BfTestAnswerElement();
+    BatfishJobExecutor<BfTestJob, BfTestAnswerElement, BfTestResult, Set<String>> executor =
+        new BatfishJobExecutor<>(
+            settings, _logger, settings.getHaltOnParseError(), TEST_EXECUTOR_DESC);
+    long startTime = System.currentTimeMillis();
+    executor.executeJobs(jobs, output, ae);
+    long elapsedTime = System.currentTimeMillis() - startTime;
+
+    // Without polling and correct parallel execution, following should be true
+    assertThat(elapsedTime, lessThanOrEqualTo(2 * TEST_JOB_DELAY));
```

Comment:
This type of test should not exist – it will be flaky. There's no guarantee that the user's machine won't be under high load, go into garbage collection, or get swapped out during the test running.

If you want to test a virtual clock, great. (E.g., inject a fake system clock and move it forward whenever the user calls sleep)

But we should not have any unit tests that insist some code finishes in a fixed wall-clock time. Instead, we should have benchmarks and a dashboard that shows how the benchmark performance changes over time.

Reply:
removed